build-tools/packages/build-cli/src/commands/generate/changelog.ts

```diff
@@ -92,16 +93,23 @@ export default class GenerateChangeLogCommand extends BaseCommand<
 		const changesetDir = path.join(releaseGroupRootDir, DEFAULT_CHANGESET_PATH);
 		const changesets = await loadChangesets(changesetDir, this.logger);
 
-		const toWrite: Promise<void>[] = [];
+		const toWriteOrDelete: Promise<void>[] = [];
 		for (const changeset of changesets) {
+			// changesets that apply to no packages should be omitted from the changelogs.
+			if (Object.keys(changeset.metadata).length === 0) {
+				this.info(`Removing changeset not included in changelog: ${changeset.sourceFile}`);
+				toWriteOrDelete.push(rm(changeset.sourceFile));
+			}
+
 			const metadata = Object.entries(changeset.metadata).map((entry) => {
 				const [packageName, bump] = entry;
 				return `"${packageName}": ${bump}`;
 			});
 			const output = `---\n${metadata.join("\n")}\n---\n\n${changeset.summary}\n\n${changeset.body}\n`;
-			toWrite.push(writeFile(changeset.sourceFile, output));
+			this.info(`Writing stripped changeset content: ${changeset.sourceFile}`);
+			toWriteOrDelete.push(writeFile(changeset.sourceFile, output));
 		}
-		await Promise.all(toWrite);
+		await Promise.all(toWriteOrDelete);
```

Comment:
I'm unclear on why we need to special case the 0 package case. Doing so adds an edge case that makes testing and code maintenance harder.

It seems like this code would work just fine without this special case.

That said, it seems off this special case is removing the file instead of writing one. I assume where is some code somewhere which removes the changesets already and I imagine it can be made to work just fine for ones that have no packages listed instead of adding a separate codepath for deleting files.


SOme more docs would help make this more understandable. The docs for stripAdditionalMetadata refer to "additional metadata" but its unclear what this metadata is in addition to.

There is also no indication of why these changes are being apply to the changesets, which makes it hard to know if deleting files with no packages is a correct implementation of "Removes any additional metadata from all changesets"


Also, probably nor practical to fix/cleanup in this PR, but generally Promise<void> functions are a bit of a code smell. Generally, functions with side-effects are harder to maintain/document/test than pure functional ones which compute a result from their inputs. Combine that with async, and you also introduce the risk that since nothing ever needs the result of an async void function, its easy to write code which expects it to be done without ever awaiting its result, it tends to be extra fragile. There are a ton of async void returning functions here, which isn't great.

If I were designing `stripAdditionalMetadata`, I'd replace the word additional with something more specific (Maybe this is stripping changelog metadata? (if so this seems wrong since you would to that to generate release notes and this is deleting items which should be in the release notes... I'm unclear on what this is for), and return the result, prabably as a string array or map rather than doing it in the file system (make it it a synchronous pure function). If I had to do it as fields for some reason, writing to a temp directory and returning a path to it would be another way to make this more robust (that would also make it not break things horribly is two async functions like this ran at once)

Reply:
Added some more comments, renamed the stripMetadata function to `canonicalizeChangesets` (it's still private and only used in the generate:changelog command), and updated the wiki page with more info about the custom metadata: https://github.com/microsoft/FluidFramework/wiki/Changesets#custom-metadata